doc/cache_layout.md

```diff
@@ -0,0 +1,243 @@
+
+## Location
+
+For the Dart SDK the default pub cache location is `$HOME/.pub_cache` on Linux
+and Mac OS, and `%LOCALAPPDATA%/Pub/Cache` on Windows.
+
+For the Flutter SDK after 3.8 uses the same folder. Before it would use the same
+folder if it was found, and otherwise  use `$FLUTTER_ROOT/.pub_cache`.
+
+The environment variable `PUB_CACHE` can be used to override the location of the
+pub cache.
+
+In this document we refer to this as `.pub_cache`.
+
+## Layout
+
+The layout of the pub cache has evolved over time, and where possible we strive
+for backwards and forwards compatibility where possible, such that a new and an
+old sdk can share the same cache.
+
+Here are the top-level folders you can find in a Pub cache.
+
+```plaintext
+.pub-cache/
+├── global_packages # Globally activated packages
+├── bin # Executables compiled from globally activated packages.
+├── git # Cloned git packages
+├── hosted # Hosted package downloads
+├── hosted-hashes # Hashes of hosted packages
+├── log # Logs after crashes and --verbose
+├── README.md # Short description of the folder
+└── _temp # Package downloads are extracted here, and moved atomically.
+```
+
+Before Dart 2.15 pub would also store credentials in the pub cache. They are now
+stored in a platform specific config dir:
+
+* On Linux $XDG_CONFIG_HOME/dart/pub-credentials.json if $XDG_CONFIG_HOME is
+  defined, otherwise $HOME/.config/dart/pub-credentials.json
+* On Mac OS: $HOME/Library/Application Support/dart/pub-credentials.json
+* On Windows: %APPDATA%/dart/pub-credentials.json
+
+### Hosted
+
+The `hosted` folder contains one folder per repository that Pub has retrieved packages from.
+
+See [hosted](../lib/src/source/hosted.dart) for details.
+
+```plaintext
+.pub-cache/hosted
+├── pub.dartlang.org
+├── pub.dev
+└── pub.flutter-io.cn
+```
+
+Before Dart 2.19 pub would by default download from `pub.dartlang.org`. This was
+changed to `pub.dev`. The two sites are mirrors and should always be identical.
+We decided to make the switch when we introduced content-hashes, because they
+anyway required redownloading of all packages to calculate the hashes.
+
+The url of the repository is encoded to a directory name with a weird URI-like
+encoding. This is a mistake that seems costly to fix, but is worth being aware
+of.
+
+Each repository folder has a sub-folder per `$package-$version` that is
+downloaded from that repository:
+
+```plaintext
+.pub-cache/hosted/pub.dev/
+├── .cache
+├── args-2.3.2
+├── retry-1.0.0
+├── yaml-3.1.1
+├── yaml_edit-2.0.2
+└── yaml_edit-2.1.0
+```
+
+A package name can always be used as a file-name (TODO: should we have a length-restriction on package-names?).
+
+A serialized version string can always be encoded as a file-name.
+
+These subfolders contain the content of the packages as they are extracted from
+the package archives. These are extracted in `.pub_cache/_temp` and moved here
+atomically, so ideally the packages here should always be fully extracted.
```

Comment:
```suggestion
atomically, hence, packages here are always fully extracted.
```
This is an invariant, and we do rely on it.

Reply:
Yeah - but we don't control the file-system. If any other process touches the files it can be broken - that's why I wrote it vaguely. But I think your wording is better...